pkg/daemon/controller/controller.go

```diff
@@ -202,6 +202,128 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
 	return ctrl.Result{}, nil
 }
 
+// Cleanup removes an instance from the cluster. If the parameter flag is true, the node object is not deleted.
+func (sc *ServiceController) Cleanup(preserveNode bool) error {
+	var node core.Node
+	err := sc.client.Get(sc.ctx, client.ObjectKey{Name: sc.nodeName}, &node)
+	if err != nil {
+		return err
+	}
+
+	desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+	if !present {
+		return errors.Errorf("node missing desired version annotation")
+	}
+	// Fetch the CM of the desired version
+	cm := &core.ConfigMap{}
+	err = sc.client.Get(sc.ctx,
+		client.ObjectKey{Namespace: sc.watchNamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	if err != nil {
+		return err
+	}
```

Comment:
This is done in multiple places, you can move this to its own function and return a custom error to identify that the annotation is missing for the logic in the next commit

Reply:
Should we log the errors and make a best effort to clean up the node? Why should we not try to clean up the instances just because the node object is missing?

Reply:
Good point. The best effort cleanup was originally only if the desired version annotation or services ConfigMap tied to it was missing. I'll extend that to occur if the node object is missing.